### ISSUE TITLE
I don't understand how to avoid using a function-scoped fixture

### ISSUE THREAD

**Issue by ckp95** [HypothesisWorks/hypothesis#2731]
TLDR: getting this warning and don't know what to do about it:

`HypothesisDeprecationWarning: tests/test_xlwings2.py::test_cos_with_floats uses the 'cos_sheet' fixture, which is reset between function calls but not between test cases generated by `@given(...)`.  You can change it to a module- or session-scoped fixture if it is safe to reuse; if not we recommend using a context manager inside your test function.  See https://docs.pytest.org/en/latest/fixture.html#sharing-test-data for details on fixture scope.`

-----

What I'm trying to do:

There is a Python library called [xlcalculator](https://github.com/bradbase/xlcalculator). It's supposed to be an implementation of the Excel calculation engine but entirely in Python and open-source, so you can integrate live spreadsheet models with Python's data science infrastructure, even in environments without an MS Office install. So, it is important that the Python implementations of formulae match the Excel implementations as closely as possible. There are a lot of undocumented quirks in Excel, which have to be discovered through experimentation.

There is another library called [xlwings](https://www.xlwings.org/). This lets you use Python to "drive" an open Excel instance, like what VBA can do. This is where hypothesis comes in. If we have an Excel formula, e.g. `COS`, and a corresponding Python function `COS`, we want to test the property that the latter matches the former's behaviour exactly (i.e. the Excel implementation is an oracle). So a test would proceed as follows:

- Launch an Excel process and open a workbook
- In the workbook, set up one cell (say, A1) to be the "input" cell, and another cell (say, B1) to be the output cell. the former would be left blank, the latter would have the formula `=COS(A1)` entered
- Use Hypothesis to generate floating point numbers. For each float `x`, enter it in cell A1, then record the new value of cell B1.
- Compare this value to the Python implementation of `COS(x)`. If they differ, fail the test.
- Once we are done with the function, clean up by closing the workbook. Make a new workbook to test a new formula, repeat.
- Once we are done with the test session, kill the Excel process

Here is code that does this (Windows only, and you need Excel installed):

```python
from contextlib import contextmanager
import math

import xlwings as xw
import pytest
from hypothesis import given, settings
from hypothesis.strategies import floats


# this is really slow so we only want to do it once per session
@pytest.fixture(scope="session")
def excel_app():
    app = xw.App(add_book=False)
    try:
        yield app
    finally:
        app.kill()


@contextmanager
def excel_workbook(app):
    wb = app.books.add()
    try:
        yield wb
    finally:
        wb.close()


# we use a different fixture for each different formula (with the boilerplate suitably abstracted away), but the fixture is reused across `hypothesis.given` examples within each `test_xxxx` function
@pytest.fixture
def cos_sheet(excel_app)
    with excel_workbook(excel_app) as wb:
        sheet = wb.sheets["Sheet1"]
        sheet["B1"].value = "=COS(A1)"
        yield sheet


# python implementation of excel formula
def COS(x):
    return math.cos(x)


@given(value=floats(allow_nan=False, allow_infinity=False))
@settings(deadline=None)
def test_cos_with_floats(value, cos_sheet):
    # arrange
    cos_sheet["A1"].value = value # insert given float into the spreadsheet
    expected = cos_sheet["B1"].value # observe the result
    
    # act
    result = COS(value)

    # assert
    assert math.isclose(result, expected)
```

(And this does actually produce a counterexample: [it fails on inputs above 2^27](https://stackoverflow.com/questions/15579897/why-does-cos227-fail), which is a limitation not mentioned in the Microsoft documentation)

But it gives the warning:

`HypothesisDeprecationWarning: tests/test_xlwings2.py::test_cos_with_floats uses the 'cos_sheet' fixture, which is reset between function calls but not between test cases generated by `@given(...)`.  You can change it to a module- or session-scoped fixture if it is safe to reuse; if not we recommend using a context manager inside your test function.  See https://docs.pytest.org/en/latest/fixture.html#sharing-test-data for details on fixture scope.`

I can't make sense of this. I *want* it to not be reset between test cases generated by `given`, because closing and making a new workbook is an expensive operation. It's totally impractical to do it anew for every single generated random float, so it has to be reused. But I *don't want* it to be module-scoped, because I don't want to share the workbooks *between* `test_` functions, because they all need a slightly different cell setup (e.g. different number of arguments, auxiliary cells to handle errors, etc). If I made it module-scoped, I'd need to make a separate little module for every single test function, and that's absurd. And I don't know what kind of context manager would help me here.

So, what do I do? I understand that this warning [is soon going to become an error](https://github.com/HypothesisWorks/hypothesis/issues/2709#issuecomment-751335212), so I need to find out the "proper" way of handling this situation.

**Comment by Zac-HD** [comment]
First of all, thanks for a well-written question!

Unfortunately I don't have a good answer for you, because you actually want the behaviour which is usually a subtle error.  The best option is probably to [fund or develop tighter pytest/Hypothesis integration](https://hypothesis.works/articles/pytest-integration-sponsorship/) which would fix the underlying problem - nobody has stepped up to do this yet, but it's probably cheaper and certainly better than working around it.

Context-manager-based workarounds will all run `workbook.close()` at the end of each test, which is semantically OK but too slow for you.  An ugly but working option might look like:

```python
def test_cos_with_floats(cos_sheet):
    @given(value=floats(allow_nan=False, allow_infinity=False))
    @settings(deadline=None)
    def inner(value):
        cos_sheet["A1"].value = value
        expected = cos_sheet["B1"].value
        assert math.isclose(COS(value), expected)

    inner()
```

Finally, maybe we should add support for e.g. `@pytest.mark.hypothesis_function_scoped_is_fine` so you can opt-out of this check; it would be occasionally useful but frequently misused 😕

**Comment by ckp95** [comment]
Thanks, the nested function way worked. Perhaps this could be added to the docs until there's a better solution?

**Comment by Zac-HD** [comment]
@ckp95 - even better, just wait for #2733 and we'll make it a health check instead of a hard error for you 😄
